### PR TITLE
Migrate DeepJIT CUDA integration to the PyTorch stable ABI

### PR DESCRIPTION
**Pull request by cleonard530**
## Summary

- Migrate CUDA stream APIs to the PyTorch 2.10 stable ABI
- Update build configuration, tests, and documentation

The Ascend build is unchanged.

Note: The CUDA extension uses PyTorch’s stable ABI, but it is not CPython ABI-stable because DeepJIT’s GIL helper still depends on pybind11 and CPython-specific APIs. Separate builds are required per Python version.

cc @janeyx99, @Harry-Chen

## Testing

- `python tests/test_cuda.py` : All DeepJIT tests passed

ran [torch-abi-audit](https://pypi.org/project/torch-abi-audit/) on minimal extension example (found here https://github.com/deepseek-ai/DeepJIT/commit/36a98fcc005eab1ab50d84fab636774c0992776c) and got   

```
[STABLE  ] [uses-private-api      ] _deep_jit_stable_abi_probe.cpython-312-x86_64-linux-gnu.so  (stable_shim=41, unstable=0)
        cpython violation: PyDict_SetDefault
        cpython violation: PyFrame_GetBack
        cpython violation: PyGILState_Check
        cpython violation: PyInstanceMethod_New
        cpython violation: PyInstanceMethod_Type
        cpython violation: _PyObject_GetDictPtr
        cpython violation: _PyThreadState_UncheckedGet
        cpython violation: _PyType_Lookup 
```
        
Most of these stable shims came in from included headers, but some came from `#include <deep_jit/backend/cuda/backend.hpp>`, and none of them were unstable. 

------------------------------------------------------------------------
## Bencmark
These benchmarks compare main against the stable-ABI branch on the SM90 DeepGEMM DeepJIT paths using the same workloads and report both `first call` and `warm call` timings. `First call` measures cold-start end-to-end latency, including JIT compilation, loading, and initial launch overhead in a fresh subprocess with a fresh JIT cache, while `warm call` measures steady-state per-launch GPU execution after warmup. For evaluating the specific deep_jit launch-path change, `warm call` is the more direct signal because it isolates runtime launch behavior from one-time compilation costs. The kernels and parameters shapes used for this benchmark are:

- `fp8_gemm_nt`:   `a shape (4096, 2048)`, `b shape (7168, 2048)`
- `grouped_fp8_nt`: `a shape (3712, 2048)`, `b shape (4, 4096, 2048)`, `grouped_layout shape (3712)`
- `tf32_hc_prenorm_gemm`: `a shape (4096, 7168)`, `b shape (24, 7168)`, `num_splits=16`
- `fp8_mqa_logits`: `seq=512` `kv=4096` `h=32` `d=128`

First Call
| Kernel | main (ms) | stable (ms) | diff (ms)
|---|---|---| --- |
| fp8_gemm_nt| 3747.508 | 3915.070 | +167.562 (4.47% slower)  
| grouped_fp8_nt | 5453.177 | 5519.743 | +66.566  (1.22% slower)
| tf32_hc_prenorm_gemm | 1610.871 | 1584.711 | -26.160 (1.62% faster) 
| fp8_mqa_logits | 1542.386 | 1482.790 | -59.596 (-3.86% faster)

Warm Calls (average over 500 calls)
| Kernel | main (μs) | stable (μs) | diff (μs)
|---|---|---| --- |
| fp8_gemm_nt| 177.9 | 177.3 | -0.6 (0.34% faster)  
| grouped_fp8_nt | 215.5 | 212.1 | -3.4  (1.58% faster)
| tf32_hc_prenorm_gemm | 62.2 | 62.3 | +0.1 (0.16% slower) 
| fp8_mqa_logits | 49.7 | 49.8 | +0.1 (0.2% slower)

Across the SM90 DeepJIT benchmark cases, the stable-ABI branch runtime is essentially the same as main.

### CODE REVIEW

File: include/deep_jit/backend/cuda/kernel.hpp [hunk 1]
```diff
@@ -29,6 +31,15 @@ inline void* kernel_arg_pointer(const T& value) {
     }
 }
 
+// Utility to get the current CUDA stream for a given device using stable APIs.
+// Returns a CUstream for use with the CUDA Driver API.
+inline CUstream get_current_cuda_stream(const int32_t device_index) {
+    void* stream_ptr = nullptr;
+    TORCH_ERROR_CODE_CHECK(
+        aoti_torch_get_current_cuda_stream(device_index, &stream_ptr));
```
Comment (janeyx99) [inline include/deep_jit/backend/cuda/kernel.hpp:39]: For more recent torch, we should have a more high level better UX API for this, no?

File: tests/test_cuda_proj/main.cpp [hunk 2]
```diff
@@ -52,6 +53,40 @@ int python_api_num_initializations = 0;
 std::shared_ptr<Runtime> process_test_runtime;
 std::shared_ptr<Runtime> gil_test_runtime;
 
+cudaStream_t get_stream_from_pool(const int32_t device_index) {
+    void* stream = nullptr;
+    TORCH_ERROR_CODE_CHECK(
+        torch_get_cuda_stream_from_pool(false, device_index, &stream));
```
Comment (janeyx99) [inline tests/test_cuda_proj/main.cpp:59]: Would be good to upstream a high level C++ wrapper API for this one

File: tests/test_cuda_proj/main.cpp [hunk 3]
```diff
@@ -52,6 +53,40 @@ int python_api_num_initializations = 0;
...
+    return static_cast<cudaStream_t>(stream);
+}
+
+class TorchCUDAStreamGuard {
+public:
+    TorchCUDAStreamGuard() = delete;
+
+    explicit TorchCUDAStreamGuard(cudaStream_t stream, int32_t device_index) {
+        TORCH_ERROR_CODE_CHECK(
+            aoti_torch_create_cuda_stream_guard(
```
Comment (janeyx99) [inline tests/test_cuda_proj/main.cpp:69]: This one too

File: tests/test_cuda.py [hunk 4]
```diff
@@ -222,6 +222,7 @@ def validate_header_self_containment(temporary_dir):
         source_path.write_text(f'#include <{header.as_posix()}>\n', encoding='utf-8')
         command = [
             os.environ.get('CXX', 'c++'), '-std=c++20', '-fsyntax-only', '-Werror',
+            '-DTORCH_TARGET_VERSION=0x020a000000000000', '-DUSE_CUDA',
```
Comment (janeyx99) [inline tests/test_cuda.py:225]: This can be 2.13

**Review by janeyx99 — COMMENTED** [review]
Thanks for looking into this! Overall I’m glad we have the shims for the APIs but we should upstream the headeronly bits for better UX.

Could you add a benchmark on a deepJIT workload to validate this change?

File: include/deep_jit/backend/cuda/kernel.hpp (hunk 1 above)
Comment (cleonard530) [reply include/deep_jit/backend/cuda/kernel.hpp]: Yes, we definitely have this for torch >= 2.13. I wasn't sure how restrictive we wanted to be here, but seeing your comment below we can update this.

**Comment by cleonard530** [comment]
> Could you add a benchmark on a deepJIT workload to validate this change?

Will do!

**Comment by cleonard530** [comment]
@janeyx99, I added the benchmark results to the description.

File: tests/test_cuda.py [hunk 5]
```diff
@@ -787,6 +788,7 @@ def run_worker():
             sources=[str(TEST_CUDA_PROJECT / 'main.cpp')],
             extra_cflags=[
                 '-std=c++20', '-O3', '-fPIC', '-Wall', '-Wextra', '-Werror',
+                '-DTORCH_TARGET_VERSION=0x020a000000000000', '-DUSE_CUDA',
```
Comment (mehmetoguzderin) [inline tests/test_cuda.py]: This does not match the README of PR, is there a particular reason for the different version here?

```suggestion
                '-DTORCH_TARGET_VERSION=0x020d000000000000', '-DUSE_CUDA',
```

File: CMakeLists.txt [hunk 6]
```diff
@@ -17,5 +17,9 @@ include_directories(
 link_directories(${TORCH_INSTALL_PREFIX}/lib ${CUDA_TOOLKIT_ROOT_DIR}/lib64 ${CUDA_TOOLKIT_ROOT_DIR}/lib64/stubs)
 
 pybind11_add_module(_C csrc/python_api.cpp)
+target_compile_definitions(_C PRIVATE
+    TORCH_TARGET_VERSION=0x020a000000000000
```
Comment (mehmetoguzderin) [inline CMakeLists.txt]: Also same here, the definition does not match README for 2.13 version.

```suggestion
    TORCH_TARGET_VERSION=0x020d000000000000
```

**Review by mehmetoguzderin — COMMENTED** [review]
Code seems to use different target version at different places, is that intentional? Thanks a lot in advance.

**Comment by cleonard530** [comment]
> Code seems to use different target version at different places, is that intentional? Thanks a lot in advance

No that was not intentional. Thanks for catching that!

**Comment by mehmetoguzderin** [comment]
Thank you @cleonard530 for the clarification & update!

File: README.md [hunk 7]
```diff
@@ -27,7 +27,7 @@ DeepJIT handles the JIT infrastructure so that kernel libraries can focus on the
...
 | **Ascend** | Bisheng and ld.lld compile and link Ascend kernel source; ACL loads and launches kernels. | CANN with `bin/bisheng`, `bin/ld.lld`, and the Ascend `adv_api` headers; ACL and `torch_npu` headers and runtime. |
 
-The host environment must provide Linux, a C++20 compiler and standard library with `std::format` support, Python, pybind11, and the dependencies for the selected backend. DeepJIT is intended to be embedded into your extension as a header-only dependency.
+The host environment must provide Linux, a C++20 compiler and standard library with `std::format` support, and the dependencies for the selected backend. The default GIL support also requires Python and pybind11; consumers that manage the GIL themselves can disable it as described under [Integration](#integration). DeepJIT is intended to be embedded into your extension as a header-only dependency. CUDA consumers should compile with `TORCH_TARGET_VERSION=0x020d000000000000` and `USE_CUDA`; this targets PyTorch's stable ABI with PyTorch 2.13 as the minimum runtime version.
```
Comment (janeyx99) [inline README.md]: @cleonard530 if we use the old cuda stream shim, what's the min runtime version we can target?

Comment (cleonard530) [reply README.md]: We could target 2.10 then (`TORCH_TARGET_VERSION=0x020a000000000000`)
